File: src/main/java/myratpackexamples/webpoll/RatpackMongoClient.java
package myratpackexamples.webpoll;

import com.mongodb.async.client.MongoCollection;
import io.vavr.control.Validation;
import lombok.Value;
import org.bson.Document;
import org.bson.conversions.Bson;
import ratpack.exec.Promise;

import java.util.ArrayList;

import static io.vavr.control.Validation.invalid;
import static io.vavr.control.Validation.valid;
import static ratpack.exec.Promise.async;

public class RatpackMongoClient {

    public static Promise<Void> insertOne(MongoCollection<Document> collection, Document document) {
        return async(downstream -> collection.insertOne(document, (result, throwable) -> {
            if (throwable != null) {
                downstream.error(throwable);
            } else {
                downstream.success(result);
            }
        }));
    }

    public abstract static class FindOneError {
    }

    @Value
    public static class ExactlyOneElementExpected extends FindOneError {
    }


    @Value
    public static class InvalidIdString extends FindOneError {
        String idString;
    }

    @Value
    public static class MongoError extends FindOneError {
        Throwable throwable;
    }

    public static Promise<Validation<FindOneError, Document>> findOne(MongoCollection<Document> collection, Bson filter) {
        return async(downstream -> collection.find(filter).into(new ArrayList<>(), (result, throwable) -> {
            if (throwable != null) {
                downstream.success(invalid(new MongoError(throwable)));
            } else {
                if (result.size() != 1) {
                    downstream.success(invalid(new ExactlyOneElementExpected()));
                } else {
                    downstream.success(valid(result.get(0)));
                }
            }
        }));
    }
}

File: src/main/java/myratpackexamples/webpoll/RatpackMongoClient.kt
package myratpackexamples.webpoll

import com.mongodb.async.client.MongoCollection
import com.mongodb.client.model.Filters
import io.vavr.control.Validation
import myratpackexamples.promises.ValidationUtil
import org.bson.Document
import org.bson.conversions.Bson
import org.bson.types.ObjectId
import ratpack.exec.Promise
import myratpackexamples.webpoll.RatpackMongoClient.FindOneError.*
import myratpackexamples.webpoll.RatpackMongoClient.InsertOneError.*

import java.util.ArrayList

import io.vavr.control.Validation.invalid
import io.vavr.control.Validation.valid
import ratpack.exec.Promise.async

object RatpackMongoClient {

    sealed class InsertOneError {
        data class InsertOneMongoError(val throwable: Throwable?): InsertOneError()
        data class InsertOneJsonProcessingError(val throwable: Throwable?): InsertOneError()
    }

    fun MongoCollection<Document>.insertOne(document: Document): Promise<Validation<InsertOneError, Void>> {
        return async { downstream ->
            insertOne(document) { result, throwable ->
                if (throwable != null) {
                    downstream.success(invalid(InsertOneMongoError(throwable)))
                } else {
                    downstream.success(valid(result))
                }
            }
        }
    }

    sealed class FindOneError {
        object ExactlyOneElementExpected: FindOneError()
        data class InvalidIdString(val idString: String?): FindOneError()
        data class FindOneMongoError(val throwable: Throwable?): FindOneError()
    }

    fun MongoCollection<Document>.findOneById(hexIdString: String?): Promise<Validation<FindOneError, Document>> {
        val mongoObjectId = createMongoObjectId(hexIdString)

        return ValidationUtil.flatMapPromise(mongoObjectId) { findOne(Filters.eq("_id", it)) }
    }

    private fun createMongoObjectId(hexIdString: String?): Validation<FindOneError, ObjectId> {
        try {
            return valid(ObjectId(hexIdString))

        } catch (e: IllegalArgumentException) {
            return invalid(InvalidIdString(hexIdString))
        }

    }

    private fun MongoCollection<Document>.findOne(filter: Bson): Promise<Validation<FindOneError, Document>> {
        return async { downstream ->
            find(filter).into(ArrayList()) { result, throwable ->
                if (throwable != null) {
                    downstream.success(invalid(FindOneMongoError(throwable)))
                } else {
                    if (result.size != 1) {
                        downstream.success(invalid(ExactlyOneElementExpected))
                    } else {
                        downstream.success(valid(result[0]))
                    }
                }
            }
        }
    }
}
